uniffi/src/ffi/rustfuture.rs

```diff
@@ -0,0 +1,586 @@
+        // Has the waker been dropped?
+        assert_eq!(Arc::strong_count(&foreign_waker), 1);
+    }
+
+    #[test]
+    fn test_wake_by_ref() {
+        let foreign_waker_environment_ptr = Box::into_raw(Box::new(RefCell::new(false)));
+        let foreign_waker_ptr =
+            RustFutureForeignWaker::new(waker as _, foreign_waker_environment_ptr as *const _)
+                .into_unit_ptr();
+        let foreign_waker = unsafe { RustFutureForeignWaker::from_unit_ptr(foreign_waker_ptr) };
+
+        // Clone to increase the strong count, so that we can see if it has not been dropped by `wake_by_ref` later.
+        let _ = unsafe { RustFutureForeignRawWaker::clone_waker(foreign_waker_ptr) };
+        assert_eq!(Arc::strong_count(&foreign_waker), 2);
+
+        // Let's call the waker by reference.
+        unsafe { RustFutureForeignRawWaker::wake_by_ref(foreign_waker_ptr) };
+
+        // Has it been called?
+        let foreign_waker_environment = unsafe { Box::from_raw(foreign_waker_environment_ptr) };
+        assert_eq!(foreign_waker_environment.take(), true);
+
+        // Has the waker not been dropped?
+        assert_eq!(Arc::strong_count(&foreign_waker), 2);
+
+        // Dropping manually to avoid data leak.
+        unsafe { RustFutureForeignRawWaker::drop_waker(foreign_waker_ptr) };
+    }
+}
+
+/// Poll a [`RustFuture`]. If the `RustFuture` is ready, the function returns
+/// `true` and puts the result inside `polled_result`, otherwise it returns `false`
+/// and _doesn't change_ the value inside `polled_result`.
+///
+/// Please see the module documentation to learn more.
+///
+/// # Panics
+///
+/// The function panics if `future` or `waker` is a NULL pointer.
+#[doc(hidden)]
+pub fn uniffi_rustfuture_poll<T>(
+    future: Option<&mut RustFuture<T>>,
+    waker: Option<NonNull<RustFutureForeignWakerFunction>>,
+    waker_environment: *const RustFutureForeignWakerEnvironment,
+    polled_result: &mut <T as FfiConverter>::FfiType,
+    call_status: &mut RustCallStatus,
+) -> bool
+where
+    T: FfiConverter,
+{
+    let future = future.expect("`future` is a null pointer");
+    let waker = waker.expect("`waker` is a null pointer");
+
+    let future_mutex = Mutex::new(future);
+
+    match call_with_output(call_status, || {
+        future_mutex
+            .lock()
+            .unwrap()
+            .poll(waker.as_ptr(), waker_environment)
+    }) {
+        Poll::Ready(result) => {
+            *polled_result = result;
+
+            true
+        }
+
+        Poll::Pending => false,
+    }
+}
+
+/// Drop a [`RustFuture`].
+///
+/// Because this function takes ownership of `future` (by `Box`ing it), the
+/// future will be dropped naturally by the compiler at the end of the function
+/// scope.
+///
+/// Please see the module documentation to learn more.
+#[doc(hidden)]
+pub fn uniffi_rustfuture_drop<T>(
+    _future: Option<Box<RustFuture<T>>>,
+    _call_status: &mut RustCallStatus,
+) where
+    T: FfiConverter,
+{
+}
```

Comment:
This seems unnecessary. The box would also be freed by the function that calls this one, if its body was empty.

Reply:
It's just to have a single place with all the declarations. The proc macros just redirect to this place. The compiler is smart enough to remove it. It also helps for the documentation and the global understanding of this. Plus, it gives a place to trace/to log to help debugging session. I prefer to keep it even if yes, it's useless as is.